sashimi-webapp/src/logic/renderer/core.js

```diff
@@ -0,0 +1,174 @@
+        imageArray[i].onerror = increateLoadedImageCount;
+      }
+      checkForLoadingCompletion();
+    });
+  },
+
+  /**
+   * Get the height of all elements inside a referenceFrame.
+   * @return {array} childHeights - array containing reference to each elements and its respective height
+   */
+  getChildHeights: function getChildHeights(referenceFrame) {
+    const childNodes = referenceFrame.childNodes;
+    const childArray = Object.keys(childNodes).map(key => childNodes[key]);
+
+    const childHeights =
+      childArray.filter(childNode => (childNode.nodeName !== '#text'))
+                .map((childNode) => {
+                  const nodeStyle = childNode.currentStyle || getComputedStyle(childNode);
+
+                  // Get node's height
+                  const nodeStyleHeight = parseFloat(nodeStyle.height, 10) || 0;
+                  const nodeHeight = Math.max(
+                    childNode.clientHeight,
+                    childNode.offsetHeight,
+                    nodeStyleHeight
+                  );
+
+                  // Get node's margin
+                  const nodeMargin = parseFloat(nodeStyle.marginTop, 10) +
+                                    parseFloat(nodeStyle.marginBottom, 10);
+
+                  const totalHeight = nodeHeight + nodeMargin;
+                  if (isNaN(totalHeight)) {
+                    throw new Error('Error calculating element\'s height');
+                  }
+
+                  return ({
+                    height: totalHeight,
+                    ele: childNode
+                  });
+                });
+
+    return childHeights;
+  },
+
+  /**
+   * This function will take in an array of elements with their heights information
+   * to organise them into an array of page according to the page size specified.
+   * @param {PageRenderer} pageRenderer
+   * @param {array} childHeights - array containing reference to each elements and its respective height
+   * @return {array} virtualBookPages - array of pages containing references to element
+   */
+  getPaginationVirtualDom: function getPaginationVirtualDom(pageRenderer, childHeights) {
+    const pr = pageRenderer;
+
+    const virtualBook = new VirtualBook();
+    let virtualPage = new VirtualPage(pr.renderHeight);
+
+    // Allocate element in pages within the render height
+    childHeights.forEach((element, index) => {
+      try {
+        virtualPage.add(element);
+      } catch (error) {
+        // Store existing page first
+        virtualBook.add(virtualPage);
+
+        if (error.message === 'Element is larger than page') {
+          if (virtualPage.filledHeight > 0) {
+            // if currently not at the beginning of page,
+            // create new page before inserting.
+            // TODO: Consider breaking element into smaller chunk
+            virtualPage = new VirtualPage(pr.renderHeight);
+          }
+          virtualPage.forceAdd(element);
```

Comment:
> "Element that has a height greater than the render height of the page will overflow out of the page.

`forceAdd` is the one that give this result. Which could be minimized by breaking large nested element (e.g. a long list of sentences) into smaller chunks.

> Some of the element height may change

As for this, I suspect, it is caused by the height not being generated correctly when it is still at the referenceFrame: `getChildHeights: function getChildHeights(referenceFrame)`.